File: Baekjoon/Problem_11000~11999/Problem_11052/Problem_11052.py
import sys

input = lambda: sys.stdin.readline().rstrip()
    
if __name__ == "__main__":
    N = int(input()) # 구입하고자하는 카트의 장수
    P = list(map(int, input().split())) # i+1장 들어있는 카드팩의 가격 P[i]
    
    max_p = [ 0 ] * (N+1)
    for i in range(1, N+1):
        max_p[i] = P[i-1]
        
        for j in range(i // 2 +1):
            max_p[i] = max(max_p[i], max_p[j] + max_p[i-j])
    
    print(max_p[-1])
    

File: Baekjoon/Problem_11000~11999/Problem_11052/Problem_11052.cpp
#include <cstdio>
#include <algorithm>

const int MAXN = 1000;

using namespace std;

int main() {
    int N; scanf("%d", &N);

    int P[MAXN]; 
    for (int i = 0; i < N; i++) scanf("%d", &P[i]);

    int maxp[MAXN+1] = { 0, };
    for (int i = 1; i <= N; i++) {
        maxp[i] = P[i-1];

        for (int j = 0; j <= i / 2; j++) {
            maxp[i] = max(maxp[i], maxp[j] + maxp[i-j]);
        }
    }

    printf("%d\n", maxp[N]);

    return 0;
}